docs/development/fleetlock/protocol.md

```diff
@@ -0,0 +1,95 @@
+---
+parent: Development
+nav_order: 3
+layout: default
+---
+
+# FleetLock protocol
+
+This document describes an HTTP-based protocol for orchestrating fleet-wide reboots, used by Zincati.
+It is modeled after a distributed counting semaphore with recursive locking and lock-ownership.
+
+## Overview
+
+This is a request-response protocol, where operations are always initiated by the client (i.e. Zincati).
+Each operation consists of a JSON payload sent as a POST request to the server.
+
+At an high level, the client can perform two operations:
+
+ * `RecursiveLock`: try to reserve (lock) a slot for rebooting
+ * `UnlockIfHeld`: try to release (unlock) a slot that it was previously holding
+
+Semaphore locks are owned, so that only the client that created a lock can release it. All operations are recursive, so that multiple unbalanced lock/unlock actions are allowed.
```

Comment:
```suggestion
Semaphore locks are owned, so that only the client that created a lock can release it. All operations are recursive meaning multiple unbalanced lock/unlock actions are allowed.
```